godel_process_path_generation/src/mesh_importer.cpp

```diff
@@ -261,28 +261,56 @@ bool MeshImporter::calculateBoundaryData(const pcl::PolygonMesh &input_mesh)
   return true;
 }
 
-void MeshImporter::computeLocalPlaneFrame(const Eigen::Hyperplane<double, 3> &plane, const Vector3d &centroid)
+void MeshImporter::computeLocalPlaneFrame(const Eigen::Hyperplane<double, 3> &plane, const Vector4d &centroid, const Cloud& cloud)
 {
-  Eigen::Vector3d origin = plane.projection(centroid);       // Project centroid onto plane
-
-  // Check if z_axis (plane normal) is closely aligned with world x_axis:
-  // If not, construct transform rotation from X,Z axes. Otherwise, use Y,Z axes.
+  Eigen::Vector3d origin = plane.projection(centroid.head(3));       // Project centroid onto plane
   const Eigen::Vector3d& plane_normal = plane.coeffs().head(3);
-  if (std::abs(plane_normal.dot(Vector3d::UnitX())) < 0.8)
-  {
-    Eigen::Vector3d x_axis = plane.projection(origin + Eigen::Vector3d::UnitX())-origin;
-    plane_frame_.matrix().col(0).head(3) = x_axis.normalized();
-    plane_frame_.matrix().col(2).head(3) = plane_normal.normalized();
-    plane_frame_.matrix().col(1).head(3) = (plane_normal.normalized().cross(x_axis.normalized())).normalized();
-  }
-  else
+
+  // Compute major axis of the part
+  Eigen::Matrix3d covar_matrix;
+  pcl::computeCovarianceMatrixNormalized (cloud, centroid, covar_matrix);
+  // Find eigenvectors
+  Eigen::Matrix3d evecs;  // eigenvectors
+  Eigen::Vector3d evals; // eigenvalues
+  pcl::eigen33 (covar_matrix, evecs, evals);
+  // Y-axis can be estimated from the 2nd eigenvector
+  Eigen::Vector3d y_axis (evecs (0, 1), evecs (1, 1), evecs (2, 1));
```

Comment:
Using `.col()` and `.row()` makes the code more readable (and less error prone)
[Eigen documentation](http://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#a58c77695de3b33405f01f2fdf3dc389d)